src/clj/rems/db/users.clj

```diff
@@ -38,14 +38,17 @@
     (or (str/blank? (:userid user))
         (str/blank? (:name user)))))
 
-(defn add-user! [user userattrs]
-  (assert user)
+(defn add-user-raw!
+  "Create or update a user given a userid and a map of raw user attributes."
+  [userid userattrs]
+  (assert userid)
   (assert userattrs)
-  (db/add-user! {:user user :userattrs (json/generate-string userattrs)}))
```

Comment:
What happens here if userattrs are not valid? Or is it possible?

Reply:
We just store the attributes. The only thing user attributes really need is the userid, which is asserted here. We do some filtering when reading the DB with `invalid-user?` above.